go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,1142 @@
+			}
+			eips, err = getEgressIPs(nodeName)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		prefixes := make([]string, 0, len(subnets)+len(eips))
+		prefixes = append(prefixes, subnets...)
+		prefixes = append(prefixes, eips...)
+		return prefixes, nil
+	}
+
+	generated := []*frrtypes.FRRConfiguration{}
+	for nodeName, frrConfigs := range nodeToFRRConfig {
+		// reset node specific information
+		selectedNetworks.hostNetworkSubnets = map[string][]string{}
+		selectedNetworks.hostSubnets = []string{}
+
+		// gather node specific information
+		for _, network := range selectedNetworks.networks {
+			selectedNetworks.hostNetworkSubnets[network], err = getPrefixes(nodeName, network)
+			if err != nil {
+				return nil, nil, err
+			}
+			selectedNetworks.hostSubnets = append(selectedNetworks.hostSubnets, selectedNetworks.hostNetworkSubnets[network]...)
+			// ordered
+			slices.Sort(selectedNetworks.hostNetworkSubnets[network])
+		}
+		// ordered
+		slices.Sort(selectedNetworks.hostSubnets)
+
+		matchedNetworks := sets.New[string]()
+		for _, frrConfig := range frrConfigs {
+			// generate FRRConfiguration for each source FRRConfiguration/node combination
+			new := c.generateFRRConfiguration(
+				ra,
+				frrConfig,
+				nodeName,
+				selectedNetworks,
+				matchedNetworks,
+			)
+			if new == nil {
+				// if we got nil, we didn't match any VRF
+				return nil, nil, fmt.Errorf("%w: FRRConfiguration %q selected for node %q has no VRF matching the RouteAdvertisements target VRF or any selected network",
+					errConfig, frrConfig.Name, nodeName)
+			}
+			generated = append(generated, new)
+		}
+		// check that we matched all the selected networks on 'auto'
+		if ra.Spec.TargetVRF == "auto" && !matchedNetworks.HasAll(selectedNetworks.networks...) {
+			return nil, nil, fmt.Errorf("%w: selected FRRConfigurations for node %q don't match all selected networks with target VRF 'auto'", errConfig, nodeName)
+		}
+	}
+
+	return generated, nads, nil
+}
+
+// generateFRRConfiguration generates a FRRConfiguration from a source for a
+// specific node. Also fills matchedNetworks with the networks that have a VRF
+// that matched any router VRF of the FRRConfiguration.
+func (c *Controller) generateFRRConfiguration(
+	ra *ratypes.RouteAdvertisements,
+	source *frrtypes.FRRConfiguration,
+	nodeName string,
+	selectedNetworks *selectedNetworks,
+	matchedNetworks sets.Set[string],
+) *frrtypes.FRRConfiguration {
+	routers := []frrtypes.Router{}
+	advertisements := sets.New(ra.Spec.Advertisements...)
+
+	// go over the source routers
+	for i, router := range source.Spec.BGP.Routers {
+
+		targetVRF := ra.Spec.TargetVRF
+		var matchedVRF, matchedNetwork string
+		var receivePrefixes, advertisePrefixes []string
+
+		// check if this router's VRF matches the target VRF
+		switch {
+		case targetVRF == "auto" && router.VRF == "":
+			// match on default network/VRF, advertise node prefixes and receive
+			// any prefix of default network.
```

Comment:
i dont understand this. targetVRF is auto, the router you are looking at in an FRR config has no VRF. How do you know this network you are trying to advertise (the network selected) is the default cluster network? Or maybe you dont, and below you only set:

```
			matchedNetwork = types.DefaultNetworkName
			advertisePrefixes = selectedNetworks.hostNetworkSubnets[matchedNetwork]
			receivePrefixes = selectedNetworks.networkSubnets[matchedNetwork]
```

that's not very explicit in the case statement itself. Can we make it easier to digest by also checking if types.DefaultNetworkName is in the selectedNetworks?

Reply:
> How do you know this network you are trying to advertise (the network selected) is the default cluster network?

Because in FRR and FRR-K8s no VRF means default VRF.

> Can we make it easier to digest by also checking if types.DefaultNetworkName is in the selectedNetworks?

I am already doing that for all the cases after the switch:
```
		if matchedVRF != router.VRF || len(advertisePrefixes) == 0 {
			// either this router VRF does not match the target VRF or we don't
			// have prefixes for it (which might be due to this RA not selecting
			// this network, but not just)
			continue
		}
```

Reply:
> Because in FRR and FRR-K8s no VRF means default VRF.

Or maybe you mean how do I know the default network is being selected in the RA? Then I guess my second answer covers this.